[GraphQL] Add Module with GraphQL of  Product Alerts

vmalyk opened the pull request:
## Description (*)
Added new module with GraphQl for Product Alerts - Price and Stock. I've added mutations:

```
type Mutation {
      subscribeProductAlertStock(productId: Int!): ProductAlert
      unsubscribeProductAlertStock(productId: Int!): Boolean
      subscribeProductAlertPrice(productId: Int!): ProductAlert
      unsubscribeProductAlertPrice(productId: Int!): Boolean
} 
```
In addition, Added fields to Customer for collect Data about active subscriptions.

### Fixed Issues (if relevant)
Fixes for [magento/community-features#315](https://github.com/magento/community-features/issues/315)

### Contribution checklist (*)
 - [X] Pull request has a meaningful description of its purpose
 - [X] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/magento2#37601: [GraphQL] Add Module with GraphQL of  Product Alerts

vmalyk committed:
Added module with GraphQL of  Product Alerts

vmalyk commented:
@magento run all tests

vmalyk committed:
Fixed typos. Fixed Collection processors and split quoting in condition.

vmalyk committed:
Fixed class names in graphql schema.

vmalyk commented:
@magento run all tests

vmalyk committed:
Rework GraphQl: Moved information about subscribed products to Customer query

vmalyk commented:
@magento run all tests

vmalyk committed:
Fixed code after run tests.

vmalyk commented:
@magento run Static Tests

torhoehn commented:
@magento run all tests

bgorski commented:
@magento run Static Tests, Magento Health Index

bgorski commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/Customer/Price.php:
```diff
@@ -0,0 +1,96 @@
+
+        /* Guest checking */
+        if (null === $customerId || 0 === $customerId) {
+            throw new GraphQlAuthorizationException(__('The current user cannot perform operations on product alerts'));
```
Please be consistent with translated sentences ending with "."

bgorski commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/Customer/Price.php:
```diff
@@ -0,0 +1,96 @@
+        $store = $context->getExtensionAttributes()->getStore();
+
+        /* Guest checking */
+        if (null === $customerId || 0 === $customerId) {
```
I think this can be simplified to 
```
if (!$customerId) {
```

bgorski commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/Customer/Stock.php:
```diff
@@ -0,0 +1,101 @@
+        $store = $context->getExtensionAttributes()->getStore();
+
+        /* Guest checking */
+        if (null === $customerId || 0 === $customerId) {
```
This can be simplified to
```
if (!$customerId) {
```
Also, please be consistent with translated sentences ending with `.`

bgorski commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/Price/Subscribe.php:
```diff
@@ -0,0 +1,83 @@
+        $store = $context->getExtensionAttributes()->getStore();
+
+        /* Guest checking */
+        if (null === $customerId || 0 === $customerId) {
```
This can be simplified to
```
if (!$customerId) {
```
Also, please be consistent with translated sentences ending with `.`

bgorski commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/Price/Subscribe.php:
```diff
@@ -0,0 +1,83 @@
+            throw new GraphQlInputException(__('The current user is currently subscribed to price alert.'));
+        }
+
+        $model->save();
```
Direct usage of `save` on an entity object is deprecated - please use a resource model instead.

bgorski commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/Price/Unsubscribe.php:
```diff
@@ -0,0 +1,79 @@
+        $store = $context->getExtensionAttributes()->getStore();
+
+        /* Guest checking */
+        if (null === $customerId || 0 === $customerId) {
```
This can be simplified to
```
if (!$customerId) {
```
Also, please be consistent with translated sentences ending with .

bgorski commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/Price/Unsubscribe.php:
```diff
@@ -0,0 +1,79 @@
+            throw new GraphQlInputException(__('The current user isn\'t subscribed to price alert.'));
+        }
+
+        $model->delete();
```
Direct usage of `delete` on an entity object is deprecated - please use a resource model instead.

bgorski commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/Price/Subscribe.php:
```diff
@@ -0,0 +1,83 @@
+                ->loadByParam();
+
+        if ($model->getId()) {
+            throw new GraphQlInputException(__('The current user is currently subscribed to price alert.'));
```
Please use GraphQlAlreadyExistsException instead.

bgorski commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/Price/Unsubscribe.php:
```diff
@@ -0,0 +1,79 @@
+                ->loadByParam();
+        
+        if (!$model->getId()) {
+            throw new GraphQlInputException(__('The current user isn\'t subscribed to price alert.'));
```
Please use GraphQlNoSuchEntityException instead.

bgorski commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/ProductResolver.php:
```diff
@@ -0,0 +1,52 @@
+        array $args = null
+    ) {
+        if (!isset($value['model'])) {
+            throw new LocalizedException(__('Missing key "model" in alert value data'));
```
This doesn't look right, please re-phrase to some more proper English like `The "model" key in the alert value data is missing.`

bgorski commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/Stock/Subscribe.php:
```diff
@@ -0,0 +1,88 @@
+        $store = $context->getExtensionAttributes()->getStore();
+
+        /* Guest checking */
+        if (null === $customerId || 0 === $customerId) {
```
This can be simplified to
```
if (!$customerId) {
```
Also, please be consistent with translated sentences ending with `.`

bgorski commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/Stock/Subscribe.php:
```diff
@@ -0,0 +1,88 @@
+            throw new GraphQlInputException(__('The current user is currently subscribed to stock alert.'));
+        }
+
+        $model->save();
```
Direct usage of `save` on an entity object is deprecated - please use a resource model instead.

bgorski commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/Stock/Subscribe.php:
```diff
@@ -0,0 +1,88 @@
+                ->loadByParam();
+
+        if ($model->getId()) {
+            throw new GraphQlInputException(__('The current user is currently subscribed to stock alert.'));
```
Please use GraphQlAlreadyExistsException instead.

bgorski commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/Stock/Unsubscribe.php:
```diff
@@ -0,0 +1,78 @@
+        $store = $context->getExtensionAttributes()->getStore();
+
+        /* Guest checking */
+        if (null === $customerId || 0 === $customerId) {
```
This can be simplified to
```
if (!$customerId) {
```
Also, please be consistent with translated sentences ending with `.`

bgorski commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/Stock/Unsubscribe.php:
```diff
@@ -0,0 +1,78 @@
+                ->loadByParam();
+        
+        if (!$model->getId()) {
+            throw new GraphQlInputException(__('The current user isn\'t subscribed to stocl alert.'));
```
Please use GraphQlNoSuchEntityException instead.

bgorski commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/Stock/Unsubscribe.php:
```diff
@@ -0,0 +1,78 @@
+            throw new GraphQlInputException(__('The current user isn\'t subscribed to stocl alert.'));
+        } 
+
+        $model->delete();
```
Direct usage of `delete` on an entity object is deprecated - please use a resource model instead.

bgorski commented on app/code/Magento/ProductAlertGraphQl/README.md:
```diff
@@ -0,0 +1,3 @@
+# ProductAlertGraphQl
```
Please follow this template when adding a readme file: https://github.com/magento/devdocs/wiki/Magento-module-README.md

bgorski commented on app/code/Magento/ProductAlertGraphQl/etc/graphql/di.xml:
```diff
@@ -0,0 +1,17 @@
+            </argument>
+        </arguments>
+    </type>
+</config>
```
missing new line at the end of the file - please add

bgorski commented on app/code/Magento/ProductAlertGraphQl/etc/module.xml:
```diff
@@ -0,0 +1,16 @@
+            <module name="Magento_ProductAlert"/>
+        </sequence>
+    </module>
+</config>
```
missing new line at the end of the file - please add

bgorski reviewed (changes requested):
@vmalyk Thank you for your contribution! I left some comments with change requets - please address them - but apart from those it looks great!

vmalyk committed:
Fixes after code review:
- Added exceptions
- Conditions for checking non-logged in customer
- README.md

vmalyk committed:
Fixes after code review for Customer data resolvers:
- Added exceptions
- Conditions for checking non-logged in customer

bgorski commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/Price/Unsubscribe.php:
```diff
@@ -0,0 +1,80 @@
+
+        /* Guest checking */
+        if (!$customerId) {
+            throw new GraphQlAuthorizationException(__('The current user cannot perform operations on product alerts'));
```
Please add `.` at the end of the sentence.

bgorski commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/ProductResolver.php:
```diff
@@ -0,0 +1,52 @@
+        array $args = null
+    ) {
+        if (!isset($value['model'])) {
+            throw new LocalizedException(__('The "model" key in the alert value data is missing'));
```
Please add `.` at the end of the sentence.

bgorski reviewed (changes requested):
Thanks for the fixes - we're getting there. I left two comments with sentences not ending with `.`. 

Apart from those, I noticed that you replaced `$model->save()` and `$model->delete()` with `$model->getResourceModel()->save($model)` and `$model->getResourceModel()->delete($model)`. This is not the correct way to do this - using resource from the model is also deprecated. You can see if a method is deprecated by checking its docblock - deprecated ones have the `@deprecated` annotation, like:
```
/**
 * Retrieve model resource
 *
 * @return \Magento\Framework\Model\ResourceModel\Db\AbstractDb
 * @deprecated 101.0.0 because resource models should be used directly
 */
public function getResource()
{
    return $this->_getResource();
}
```
The correct way is to use dependency injection to pull the resource model in the class constructor and use it directly. Please do that in all 4 instances where you use `$model->getResourceModel()`.

vmalyk committed:
Call ResourceModel directly and inject model to constructors. Fixed "." in some messages.

bgorski commented:
@magento run all tests

bgorski reviewed (changes requested):
Thanks for all the fixes! I ran all the automated tests, though, and found out that there's a bit more to fix. From what I see functional tests are not failing because of your contribution and you can also ignore health index and static integrity check (those two will be handled by the internal Adobe team during the delivery), but there are some other static tests that fail. You can click on Details on the right side of Static Tests, then on Console Error Logs and then on PHP. All of those have to be fixed and then we can proceed with processing of the PR. Thanks!

vmalyk committed:
Static tests fixes: Code styling and unused private variables.

vmalyk committed:
Static tests fixes:Fixes for reduant dependencies and Composer

vmalyk commented:
@magento run all tests

vmalyk committed:
Static tests fixes:Fixes for reduant dependencies in composer.json.

vmalyk commented:
@magento run Static Tests

bgorski commented:
@magento run all tests

Den4ik commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/Price/Subscribe.php:
```diff
@@ -0,0 +1,93 @@
+                ->loadByParam();
+
+        if ($model->getId()) {
+            throw new GraphQlAlreadyExistsException(__('The current user is currently subscribed to price alert.'));
```
Are you sure that we should throw exception instead returning exist subscription data?

bgorski replied:
Throwing the exception makes sense for me. This is a clear signal for the frontend that the subscription already exists. Email newsletter subscription, for example, is doing the same thing.

vmalyk replied:
I think adding  Exception instead irrelevant saves to database. Ofc, ProductAlert module has't that check and saves alert on each request. Exception provides more flexible way  for API and provides additional options  and implementations for PWA.

Den4ik commented on app/code/Magento/ProductAlertGraphQl/Model/Resolver/Stock/Subscribe.php:
```diff
@@ -0,0 +1,98 @@
+                ->loadByParam();
+
+        if ($model->getId()) {
+            throw new GraphQlAlreadyExistsException(__('The current user is currently subscribed to stock alert.'));
```
Are you sure that we should throw exception instead returning exist subscription data?

bgorski replied:
Throwing the exception makes sense for me. This is a clear signal for the frontend that the subscription already exists. Email newsletter subscription, for example, is doing the same thing.

vmalyk replied:
I think adding  Exception instead irrelevant saves to database. Ofc, ProductAlert module has't that check and saves alert on each request. Exception provides more flexible way  for API and provides additional options  and implementations for PWA.

Den4ik reviewed (changes requested):
@vmalyk Thanks for work on this PR. Please look at my comments

bgorski reviewed (changes requested):
@vmalyk I noticed that there are still some PHP errors:
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/31467/ac924f9b35a27b494454a0f62a8b54b7/Statics/console-error-logs.html

Can you fix those?

vmalyk committed:
Fixes after code review for Customer data resolvers: Fixed warnings about line length in PHP PSR-2 tests.

vmalyk commented:
@magento run all tests

vmalyk commented:
@bgorski , thank you for your review. I've double-checked tests and now I've 2 unfixed things and look like it's unavailable for fix:

vmalyk committed:
Fixed static tests: PHPCPD fixes. Adjust README.md.

vmalyk committed:
Merge branch '2.4-develop' of github.com:magento/magento2 into add-product-alert-graph-ql

# Conflicts:
#	dev/tests/static/testsuite/Magento/Test/Php/_files/phpcpd/blacklist/common.txt

vmalyk committed:
Fixed static tests: PHPCPD fixes after merge develop.

vmalyk commented:
@magento run all tests

vmalyk committed:
Added minimal PHP requirements to composer.json as in main file.

vmalyk commented:
@magento run all tests

vmalyk commented:
@bgorski , I've just pushed changes for Composer. so, now it looks correct?

bgorski reviewed (approved):
Looks good now - appoved. Although some test types fail, it's not caused by this PR. Thanks for your contribution @vmalyk!

engcom-Alfa commented:
@magento crete issue

bgorski commented:
@magento run Static Tests, Magento Health Index, Functional Tests CE, Database Compare

Den4ik committed:
Merge branch '2.4-develop' into add-product-alert-graph-ql

Den4ik commented:
@magento run all tests

Den4ik committed:
Update php required versions

Den4ik commented:
@magento run all tests

engcom-Lima committed:
Merge branch '2.4-develop' into add-product-alert-graph-ql

engcom-Lima commented:
@magento run all tests

engcom-Lima commented:
@magento create issue

engcom-Lima commented:
Hi @vmalyk ,

Thanks for your contribution and collaboration.

Please provide us information on below points:
1. Mutation/payload.
2. Usage of mutation(flow of the mutation).
3. Backend configuration required.
4. Affected module (if any)

Thanks

engcom-Lima commented:
Hi @vmalyk ;

Please provide the information asked in this [comment](https://github.com/magento/magento2/pull/31467#issuecomment-1587345467).

Thanks

engcom-Lima committed:
Merge branch '2.4-develop' into add-product-alert-graph-ql

engcom-Lima commented:
@magento run all tests

engcom-Charlie commented:
Hi @vmalyk,

Thank you for your contribution!

Can you please provide some input on [this](https://github.com/magento/magento2/pull/31467#issuecomment-1587345467).  We have tried to run these mutation but its giving us error.

Could please share the mutation with proper input in order to work on this PR further.

Thank you!

vmalyk commented:
Payloads for testing each mutation:

engcom-Lima committed:
Merge branch '2.4-develop' into add-product-alert-graph-ql

engcom-Lima commented:
@magento run all tests

engcom-Lima commented:
Hi @vmalyk ,

Thanks for your contribution and collaboration.

As per the above [comment](https://github.com/magento/magento2/pull/31467#issuecomment-1607195543), I have used the subscription mutation but it is giving me syntax error `"Cannot query field \"subscribeProductAlertStock\" on type \"Mutation\".",`. Please refer to the below screenshots for your reference:
<img width="1418" alt="Screenshot 2023-06-27 at 1 39 24 PM" src="https://github.com/magento/magento2/assets/60198022/051c4e2c-e0ae-48cc-9880-d1dcb5dccc6b">
<img width="1452" alt="Screenshot 2023-06-27 at 1 39 54 PM" src="https://github.com/magento/magento2/assets/60198022/c609006d-339c-48da-b0c1-5a838d3854da">

Could you provide us with more information if we are missing anything?

Thanks

engcom-Lima commented:
Hi @vmalyk ,

Kindly provide us information asked in this [comment.](https://github.com/magento/magento2/pull/31467#issuecomment-1609226759)

Thanks

engcom-Lima committed:
Merge branch '2.4-develop' into add-product-alert-graph-ql

engcom-Lima commented:
@magento run all tests

engcom-Lima commented:
Hi @vmalyk ,

Thanks for your contribution and collaboration.

We await the reply to this [comment](https://github.com/magento/magento2/pull/31467#issuecomment-1609226759). Since it was not updated for a long time, we are closing this PR. Please feel free to reopen this.

Thanks
